kvbc/include/categorization/updates.h

```diff
@@ -0,0 +1,180 @@
+      // Log warn
+      return;
+    }
+    data_.deletes.emplace_back(std::move(key));
+  }
+
+  void calculateHash(const bool hash) { data_.calculate_hash = hash; }
+
+ private:
+  KeyValueUpdatesData data_;
+  std::set<std::string> unique_deletes_;
+  friend struct Updates;
+};
+
+// Updates for a merkle tree category.
+// Persists key-values in a merkle tree that is constructed on top of the underlying key-value store.
+struct MerkleUpdates {
+  MerkleUpdates() {}
+  MerkleUpdates(MerkleUpdates&& other) = default;
+  MerkleUpdates& operator=(MerkleUpdates&& other) = default;
+
+  // Do not allow copy
+  MerkleUpdates(MerkleUpdates& other) = delete;
+  MerkleUpdates& operator=(MerkleUpdates& other) = delete;
+
+  void addUpdate(std::string&& key, std::string&& val) { data_.kv.emplace(std::move(key), std::move(val)); }
+
+  void addDelete(std::string&& key) {
+    if (const auto [itr, inserted] = unique_deletes_.insert(key); !inserted) {
+      *itr;  // disable unused variable
+      // Log warn
+      return;
+    }
+    data_.deletes.emplace_back(std::move(key));
+  }
+
+  const MerkleUpdatesData& getData() { return data_; }
+
+ private:
+  MerkleUpdatesData data_;
+  std::set<std::string> unique_deletes_;
+  friend struct Updates;
+};
+
+// Updates contains a list of updates for different categories.
+// Note: Only a single `SharedKeyValueUpdates` is supported per block.
+struct Updates {
+  void addMerkleUpdate(std::string category_id, MerkleUpdates&& updates) {
+    if (const auto [itr, inserted] = categoryUpdates_.insert({category_id, std::move(updates.data_)}); inserted) {
+      *itr;  // disable unused variable
+      throw std::logic_error{std::string("Only one update for category is allowed. type: Merkle, category: ") +
+                             category_id};
+    }
+  }
+
+  void addKeyValueUpdates(std::string category_id, KeyValueUpdates&& updates) {
+    if (const auto [itr, inserted] = categoryUpdates_.insert({category_id, std::move(updates.data_)}); inserted) {
+      *itr;  // disable unused variable
+      throw std::logic_error{std::string("Only one update for category is allowed. type: KVHash, category: ") +
+                             category_id};
+    }
+  }
+
+  void addSharedKeyValueUpdates(std::string category_id, SharedKeyValueUpdates&& updates) {
+    if (sharedUpdate_.has_value()) {
+      throw std::logic_error{std::string("Only one update for category is allowed. type: Shared KV, category: ") +
+                             category_id};
+    }
+    sharedUpdate_.emplace(std::move(updates.data_));
+  }
+
+  std::map<std::string, std::variant<MerkleUpdatesData, KeyValueUpdatesData>>& getCategoryUpdate() {
+    return categoryUpdates_;
+  }
+
+  std::optional<SharedUpdatesData>& getSharedUpdates() { return sharedUpdate_; }
+
+ private:
+  std::optional<SharedUpdatesData> sharedUpdate_;
+  std::map<std::string, std::variant<MerkleUpdatesData, KeyValueUpdatesData>> categoryUpdates_;
```

Comment:
I changed ```Updates``` to use the CMF data structs directly (previously was the wrapper structs).
Did we had a reason not to use it like this @dartdart26  ?

Reply:
I don't think we had a reason. Might just be an artefact from the previous PR.